lib/src/Pos/Util/UserPublic.hs

```diff
@@ -0,0 +1,287 @@
+import           Test.QuickCheck (Arbitrary (..))
+import           Test.QuickCheck.Arbitrary.Generic (genericArbitrary, genericShrink)
+import           Universum
+
+import           Pos.Binary.Class (Bi (..), Cons (..), Field (..), decodeFull', deriveSimpleBi,
+                                   encodeListLen, enforceSize, serialize')
+import           Pos.Binary.Crypto ()
+import           Pos.Core (Address, addressF, makeRootPubKeyAddress)
+import           Pos.Crypto (PublicKey, PublicKey)
+
+import           Test.Pos.Crypto.Arbitrary ()
+
+#ifdef POSIX
+import           Formatting (oct, sformat)
+import qualified System.Posix.Files as PSX
+import qualified System.Posix.Types as PSX (FileMode)
+import           System.Wlog (logWarning)
+#endif
+
+-- Because of the Formatting import
+{-# ANN module ("HLint: ignore Use fewer imports" :: Text) #-}
+
+-- | Describes HD wallets keyfile content
+data WalletUserPublic = WalletUserPublic
+    { _wupKey        :: PublicKey           -- ^ root key of wallet set
+    , _wupWalletName :: Text                -- ^ name of wallet
+    , _wupAccounts   :: [(Word32, Text)]    -- ^ accounts coordinates and names
+    , _wupAddrs      :: [(Word32, Word32)]  -- ^ addresses coordinates
+    } deriving (Show, Generic)
+
+deriving instance Eq PublicKey => Eq WalletUserPublic
+
+instance Arbitrary WalletUserPublic where
+    arbitrary = genericArbitrary
+    shrink = genericShrink
+
+makeLenses ''WalletUserPublic
+
+instance Buildable WalletUserPublic where
+    build WalletUserPublic{..} =
+        bprint ("{ root address = "%addressF%", name = "%build%
+                ", accounts = "%pairsF%", addresses = "%pairsF%" }")
+        (makeRootPubKeyAddress _wupKey)
+        _wupWalletName
+        _wupAccounts
+        _wupAddrs
+      where
+        pairsF :: (Buildable a, Buildable b) => Format r ([(a, b)] -> r)
+        pairsF = later $ mconcat . map (uncurry $ bprint ("("%build%", "%build%")"))
+
+deriveSimpleBi ''WalletUserPublic [
+    Cons 'WalletUserPublic [
+        Field [| _wupKey        :: PublicKey          |],
+        Field [| _wupWalletName :: Text               |],
+        Field [| _wupAccounts   :: [(Word32, Text)]   |],
+        Field [| _wupAddrs      :: [(Word32, Word32)] |]
+    ]]
+
+-- | User public data.
+data UserPublic = UserPublic
+    { _upKeys    :: [PublicKey]
+    , _upWallet  :: Maybe WalletUserPublic
+    , _upPath    :: FilePath
+    , _upLock    :: Maybe FileLock
+    } deriving (Generic)
+
+deriving instance Eq PublicKey => Eq UserPublic
+
+instance Arbitrary (Maybe FileLock) => Arbitrary UserPublic where
+    arbitrary = genericArbitrary
+    shrink = genericShrink
+
+makeLenses ''UserPublic
+
+class HasUserPublic ctx where
+    -- if you're going to mock this TVar, look how it's done for peer state.
+    userPublic :: Lens' ctx (TVar UserPublic)
+
+-- | Show instance to be able to include it into NodeParams
+instance Bi Address => Show UserPublic where
```

Comment:
Fully concrete constraint issue. Fortunately, we recently reuiniting the `Bi Address` orphan with the `Address` type, so having `Address` in scope will also have this instance in scope. We can safely delete it.